db/db_batch.go

```diff
@@ -124,121 +145,72 @@ func (b *baseKVStoreBatch) Clear() error {
 	return nil
 }
 
-// Commit needs to be implemented by derived class
-func (b *baseKVStoreBatch) Commit() error {
-	panic("not implement")
-}
-
-// KVStore needs to be implemented by derived class
-func (b *baseKVStoreBatch) KVStore() KVStore {
-	panic("not implement")
-}
-
 //======================================
-// memKVStoreBatch is the in-memory implementation of KVStoreBatch for testing purpose
+// CachedBatch implementation
 //======================================
-type memKVStoreBatch struct {
-	baseKVStoreBatch
-	kv *memKVStore
-}
-
-// NewMemKVStoreBatch instantiates an in-memory KV store
-func NewMemKVStoreBatch(kv *memKVStore) KVStoreBatch {
-	return &memKVStoreBatch{kv: kv}
-}
-
-// Commit persists pending writes to DB and clears the write queue
-func (m *memKVStoreBatch) Commit() error {
-	for _, write := range m.writeQueue {
-		if write.writeType == Put {
-			if err := m.kv.Put(write.namespace, write.key, write.value); err != nil {
-				return err
-			}
-		} else if write.writeType == PutIfNotExists {
-			if err := m.kv.PutIfNotExists(write.namespace, write.key, write.value); err != nil {
-				return err
-			}
-		} else if write.writeType == Delete {
-			if err := m.kv.Delete(write.namespace, write.key); err != nil {
-				return err
-			}
-		}
+
+// NewCachedBatch returns a new cached batch buffer
+func NewCachedBatch() CachedBatch {
+	return &cachedBatch{
+		cache: make(map[hash.PKHash][]byte),
 	}
-	// clear queues
-	return m.Clear()
 }
 
-// KVStore returns the underlying KVStore
-func (m *memKVStoreBatch) KVStore() KVStore {
-	return m.kv
+// Put inserts a <key, value> record
+func (cb *cachedBatch) Put(namespace string, key, value []byte, errorFormat string, errorArgs ...interface{}) error {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+	cb.cache[cb.hash(namespace, key)] = value
```

Comment:
we could create a private function put, which includes line 163 to line 166. Then both Put and PutIfNotExists could use it

Reply:
added batch() above